llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -772,6 +773,66 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
           unsigned CurrentOffset = Ins.first->second;
           if (IsNewSpecConstant) {
             unsigned Size = M.getDataLayout().getTypeStoreSize(SCTy);
+            unsigned Align = M.getDataLayout().getABITypeAlignment(SCTy);
+
+            // Ensure correct alignment
+            if (CurrentOffset % Align != 0) {
+              // Compute necessary padding to correctly align the constant.
+              Padding = Size - CurrentOffset % Align;
+
+              // Update offsets.
+              NextOffset += Padding;
+              CurrentOffset += Padding;
+              OffsetMap[SymID] = NextOffset;
+
+              // The spec constant map can't be empty as the first offset is 0
+              // and so it can't be misaligned.
+              assert(!SCMetadata.empty() &&
+                     "Cannot add padding to first spec constant");
+
+              // To communicate the padding to the runtime, update the metadata
+              // node of the previous spec constant to append a padding node. It
+              // can't be added in front of the current spec constant, as doing
+              // so would require the spec constant node to have a non-zero
+              // CompositeOffset which breaks accessing it in the runtime.
```

Comment:
What part of the runtime makes this restriction? I am not opposed to appending it to the previous entry, but it would be a little cleaner to have it attached to the entry forcing the alignment.

Reply:
If there's two metadata nodes and the second one has a `CompositeOffset` of `0`, the `OffsetFromLast` here goes negative and it causes issues:

* https://github.com/intel/llvm/blob/sycl/sycl/source/detail/device_image_impl.hpp#L282

And if we set the `CompositeOffset` correctly, then reading/writing the data won't work properly for emulated spec constants as it offsets the input data with it, see here:

* https://github.com/intel/llvm/blob/sycl/sycl/source/detail/device_image_impl.hpp#L131

So appending the padding works because it can have a `CompositeOffset` as it's not accessed, but putting it in front doesn't because then the data node needs a non-null `CompositeOffset`.